### PR TITLE
DecisionTreeFactor `prune` method

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR adds a new prune method to `DecisionTreeFactor` which will allow for managing exponential growth.

Given a `DecisionTreeFactor`
```
DecisionTreeFactor:
 f[ (1,2), (2,2), (3,2), ]
 Choice(3)
 0 Choice(2)
 0 0 Choice(1)
 0 0 0 Leaf    1
 0 0 1 Leaf    2
 0 1 Choice(1)
 0 1 0 Leaf    3
 0 1 1 Leaf    4
 1 Choice(2)
 1 0 Choice(1)
 1 0 0 Leaf    5
 1 0 1 Leaf    6
 1 1 Choice(1)
 1 1 0 Leaf    7
 1 1 1 Leaf    8
```

if we set `maxNrLeaves` to 5, which means only keep the top 5 leaves, we get
```
DecisionTreeFactor:
 f[ (1,2), (2,2), (3,2), ]
 Choice(3)
 0 Choice(2)
 0 0 Leaf    0
 0 1 Choice(1)
 0 1 0 Leaf    0
 0 1 1 Leaf    4
 1 Choice(2)
 1 0 Choice(1)
 1 0 0 Leaf    5
 1 0 1 Leaf    6
 1 1 Choice(1)
 1 1 0 Leaf    7
 1 1 1 Leaf    8
```
as expected.

### CODE REVIEW

File: gtsam/discrete/DecisionTreeFactor.cpp [hunk 1]
```diff
@@ -286,5 +286,43 @@ namespace gtsam {
         AlgebraicDecisionTree<Key>(keys, table),
         cardinalities_(keys.cardinalities()) {}
 
+  /* ************************************************************************ */
+  DecisionTreeFactor DecisionTreeFactor::prune(size_t maxNrLeaves) const {
+    const size_t N = maxNrLeaves;
+
+    // Let's assume that the structure of the last discrete density will be the
```
Comment (dellaert) [inline gtsam/discrete/DecisionTreeFactor.cpp:293]: stray comment?

File: gtsam/discrete/DecisionTreeFactor.cpp [hunk 2]
```diff
@@ -286,5 +286,43 @@ namespace gtsam {
...
+        return 0.0;
+      } else {
+        total += 1;
+        return value;
+      }
+    };
+    DecisionTree<Key, double> thresholded(*this, thresholdFunc);
+
+    // Create pruned decision tree factor
+    DecisionTreeFactor prunedDiscreteFactor(this->discreteKeys(), thresholded);
```
Comment (dellaert) [inline gtsam/discrete/DecisionTreeFactor.cpp:322]: combine both statements

**Review by dellaert — APPROVED** [review]
cool

**Comment by varunagrawal** [comment]
Done and done

**Pull request merged**